### ISSUE TITLE
Moving tag breaks annotated tags

### ISSUE THREAD

**Issue by lilyball** [brotherbard/gitx#38]
If I drag a tag from one commit to another, it turns the annotated tag into a lightweight tag. Instead it should try and preserve as much of the tag as possible. You can do this by using `git cat-file tag SHA1` to get the contents of the tag, change the first line to reference the new "object" that you're tagging, and pipe that to `git mktag` to create the new tag object. Then it's just a matter of updating the tag ref to point to your new tag. The only time this should fail is if the tag was signed, and even then only the signature itself is invalid.

**Comment by brotherbard** [comment]
I've been half tempted to not allow moving tags (and remote branches). Tags were created to represent a fixed point in time.

Is there a particular use case/workflow where you need to move an annotated tag?

**Comment by lilyball** [comment]
It's rare, I'll grant you that. In this case, we just did a fairly major reorg to move to [nvie's branching model](http://nvie.com/git-model), and as part of that we moved our product tags onto the new master branch. The trees were the same, and in fact the old commit referenced by the tag was the second parent of the new commit. My first attempt was to simply move a tag using GitX, but it converted it to a non-annotated tag and I had to start over. I ended up using a script that did what I suggested in the ticket.
